File: ios/SplitBackAPI/Sources/SplitBackAPI/Models/ExpenseItem.swift
import Foundation
import SwiftData

/// A line item on an expense (e.g. a single product on a receipt). Mirrors `expense_items`.
@Model
final class ExpenseItem {
    @Attribute(.unique) var id: UUID
    var name: String
    var quantity: Decimal
    var price: Decimal
    var category: String?
    /// The participant this item is assigned to for budget attribution (nil = shared/split).
    var ownerIdentifier: String?
    /// Provenance: who added/edited this item and when.
    var addedBy: String?
    var editedBy: String?
    var addedOn: Date?
    var editedOn: Date?
    var expense: Expense?

    init(
        id: UUID,
        name: String,
        quantity: Decimal,
        price: Decimal,
        category: String? = nil,
        ownerIdentifier: String? = nil,
        addedBy: String? = nil,
        editedBy: String? = nil,
        addedOn: Date? = nil,
        editedOn: Date? = nil,
        expense: Expense? = nil
    ) {
        self.id = id
        self.name = name
        self.quantity = quantity
        self.price = price
        self.category = category
        self.ownerIdentifier = ownerIdentifier
        self.addedBy = addedBy
        self.editedBy = editedBy
        self.addedOn = addedOn
        self.editedOn = editedOn
        self.expense = expense
    }
}

File: backend/app/models/expense_item.py
import uuid
from decimal import Decimal

from sqlalchemy import ForeignKey, Numeric, String
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db import Base
from app.models.base import TimestampMixin, UUIDMixin


class ExpenseItem(UUIDMixin, TimestampMixin, Base):
    __tablename__ = "expense_items"

    expense_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("expenses.id", ondelete="CASCADE"), nullable=False
    )
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    quantity: Mapped[Decimal] = mapped_column(Numeric(10, 3), nullable=False, default=1)
    price: Mapped[Decimal] = mapped_column(Numeric(12, 2), nullable=False)
    category: Mapped[str | None] = mapped_column(String(128), nullable=True)
    # Budget attribution: the participant this item is assigned to (null = shared/split).
    owner_identifier: Mapped[str | None] = mapped_column(String(255), nullable=True)
    # Provenance (added-by / edited-by; added-on / edited-on are created_at / updated_at).
    created_by: Mapped[str | None] = mapped_column(String(255), nullable=True)
    updated_by: Mapped[str | None] = mapped_column(String(255), nullable=True)

    expense: Mapped["Expense"] = relationship(back_populates="items")  # noqa: F821
